### PR TITLE
machine/samd21: Implement reset processor 

### PR DESCRIPTION
**Pull request by deadprogram**
This PR is a draft for adding a software-based system reset to be able to flash new software on SAMD21 boards without pushing the reset button to start the bootloader.

### CODE REVIEW

File: src/machine/machine_atsamd21.go [hunk 1]
```diff
@@ -2082,3 +2083,29 @@ func setEPINTENSET(ep uint32, val uint8) {
...
+	arm.DisableInterrupts()
+
+	for !sam.NVMCTRL.INTFLAG.HasBits(sam.NVMCTRL_INTFLAG_READY) {
+	}
+
+	//NVMCTRL->STATUS.reg |= NVMCTRL_STATUS_MASK;
+	sam.NVMCTRL.STATUS.SetBits(0x111)
+
+	//NVMCTRL->ADDR.reg  = (uintptr_t)&NVM_MEMORY[APP_START / 4];
+	sam.NVMCTRL.ADDR.Set(uint32(uintptr(0x00002004))) // is this correct?
```
Comment (deadprogram) [inline src/machine/machine_atsamd21.go]: This line right here is the one I am wanting to confirm, before I accidentally brick one of my boards...

Comment (deadprogram) [reply src/machine/machine_atsamd21.go]: This is the line in the linker script that shows the starting address: https://github.com/tinygo-org/tinygo/blob/master/targets/atsamd21.ld#L4

**Comment by aykevl** [comment]
The only comment I have is bikeshedding: I would have called this just `machine.Reset()` instead of `machine.ResetProcessor()` (what would you reset other than the processor?). But I have no strong feelings about it.

What do you think? Any preference?

**Comment by deadprogram** [comment]
I think that `Reset()` might be ambigous. Perhaps even better to name it `ResetToBootloader()` since it does more than just what `arm.SystemReset()` already will do.

**Comment by deadprogram** [comment]
Due to the PR #544 that uses code in THIS PR, I think it would be easier to not rename this right now, but instead after both PR's have landed I can submit a PR that renames this method, if we still think is needed.

**Comment by aykevl** [comment]
Ok, let's keep it at `machine.ResetProcessor()` then.

**Pull request merged**